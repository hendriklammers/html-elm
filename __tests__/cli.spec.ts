import path from 'path'
import { exec } from 'child_process'

const cli = (
  args: string[],
  cwd: string
): Promise<{
  code: number
  error: Error | null
  stdout: string
  stderr: string
}> =>
  // TODO: Use util.promisify
  new Promise(resolve => {
    exec(
      `ts-node ${path.resolve('./src/cli')} ${args.join(' ')}`,
      { cwd },
      (error, stdout, stderr) => {
        resolve({
          code: error && error.code ? error.code : 0,
          error,
          stdout,
          stderr,
        })
      }
    )
  })

describe('CLI', () => {
  it('should have a --help flag', async () => {
    const { stdout } = await cli(['--help'], '.')
    expect(stdout).toMatchSnapshot()
  })

  it('should convert html to elm', async () => {
    const html = `'<div>Hello world</div>'`
    const { stdout } = await cli([html], '.')
    expect(stdout).toMatchSnapshot()
  })

  it('should support all available options', async () => {
    const html = `'<section><h1 class="title">Hello</h1><div><svg width="600" height="400" viewBox="0 0 600 400"><rect width="600" height="400" fill="#FFDA1A" /></svg></div></section>'`
    const args = [
      html,
      '-i 2',
      '--imports true',
      '-t Html',
      '-a H',
      '-s Svg',
      '-g S',
    ]
    const { stdout } = await cli(args, '.')
    expect(stdout).toMatchSnapshot()
  })
})
