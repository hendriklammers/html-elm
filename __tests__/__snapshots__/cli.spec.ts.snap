// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CLI should convert html to elm 1`] = `
"div
    []
    [ text "Hello world" ]
"
`;

exports[`CLI should support all available options 1`] = `
"import Html
import Html.Attributes as H
import Svg
import Svg.Attributes as S

Html.section
  []
  [ Html.h1
    [ H.class "title" ]
    [ Html.text "Hello" ]
  , Html.div
    []
    [ Svg.svg
      [ S.width "600", S.height "400", S.viewBox "0 0 600 400" ]
      [ Svg.rect
        [ S.width "600", S.height "400", S.fill "#FFDA1A" ]
        []
      ]
    ]
  ]
"
`;
